### PR TITLE
ENH: GH34946 Check type of names argument to `read_csv`, `read_table`…

### LINKED ISSUE #34946: BUG: Inconsistent behavior: Occasional crash on same code and data
[x] I have checked that this issue has not already been reported.
[x]  I have confirmed this bug exists on the latest version of pandas.

**Description**
When reading csv with pd.read_csv, giving both names= and dtypes= arguments, the program occasionally crashes. Surprisingly, sometimes the application completes its execution cycle normally - without any change in code or data.

I had initially used only the _dtype_ argument, but the first row of my data would be considered as column headers (the csv, otherwise has no headers). Instead of using headers=None, I added _names_ argument, which on one hand made all my rows read as data while on the other set the desired column headers in DataFrame.

But after this change, the program's behavior has become inconsistent; occasional crashing while normal execution at other times.

### Reproducing Code Example
**Example Code and Data**
Minimal code along with a small data file, that demonstrate the reported behavior, is:
 
```python
import pandas as pd

df = pd.read_csv('data\\test3.csv', names={'pVal1', 'pClass'}, dtype={'pVal1':int, 'pClass':bool})
print (df)
```
The data file test3.csv contains:
```
1,TRUE
2,FALSE
3,TRUE
```

The problem goes away if we remove either of the two arguments (names, dtype). With any of the following statements, the program executes normally:
```
df = pd.read_csv('data\\test3.csv', names={'pVal1', 'pClass'})
```
or
```
df = pd.read_csv('data\\test3.csv', dtype={'pVal1':int, 'pClass':bool})
```

### Error message:
**Demonstration of Inconsistent Behaviour**

I am posting the result of two consecutive executions; first successful, the second failing with error:

**Successful Execution**
(ml) C:\Users\pyuser\projs\ml>py err.py
   pVal1  pClass
0      1    True
1      2   False
2      3    True

**Crash with Error**
(ml) C:\Users\pyuser\projs\ml>py err.py
Traceback (most recent call last):
  File "pandas\_libs\parsers.pyx", line 1152, in pandas._libs.parsers.TextReader
._convert_tokens
TypeError: Cannot cast array from dtype('int64') to dtype('bool') according to t
he rule 'safe'

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "err.py", line 3, in <module>
    df = pd.read_csv('data\\test3.csv', names={'pVal1', 'pClass'}, dtype={'pVal1
':int, 'pClass':bool})
  File "C:\Users\pyuser\projs\ml\lib\site-packages\pandas\io\parsers.py", line 676
, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "C:\Users\pyuser\projs\ml\lib\site-packages\pandas\io\parsers.py", line 454
, in _read
    data = parser.read(nrows)
  File "C:\Users\pyuser\projs\ml\lib\site-packages\pandas\io\parsers.py", line 113
3, in read
    ret = self._engine.read(nrows)
  File "C:\Users\pyuser\projs\ml\lib\site-packages\pandas\io\parsers.py", line 203
7, in read
    data = self._reader.read(nrows)
  File "pandas\_libs\parsers.pyx", line 860, in pandas._libs.parsers.TextReader.
read
  File "pandas\_libs\parsers.pyx", line 875, in pandas._libs.parsers.TextReader.
_read_low_memory
  File "pandas\_libs\parsers.pyx", line 952, in pandas._libs.parsers.TextReader.
_read_rows
  File "pandas\_libs\parsers.pyx", line 1084, in pandas._libs.parsers.TextReader
._convert_column_data
  File "pandas\_libs\parsers.pyx", line 1160, in pandas._libs.parsers.TextReader
._convert_tokens
ValueError: cannot safely convert passed user dtype of bool for int64 dtyped dat
a in column 0

**Software and Library Versions**
### Numpy/Python version information:
(ml) C:\Users\pyuser\projs\ml>python --version
Python 3.8.2

(ml) C:\Users\pyuser\projs\ml>python -c "import numpy; print(numpy.version.version
)"
1.18.4

(ml) C:\Users\pyuser\projs\ml>python -c "import pandas; print(pandas.__version__)"
1.0.5

1.18.4 3.8.2 (tags/v3.8.2:7b3ab59, Feb 25 2020, 23:03:10) [MSC v.1916 64 bit (AM
D64)]

### PR DESCRIPTION
**Pull request by MJafarMashhadi**
- [x] closes #34946 
- [x] tests passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

- ~~One test failed (TestTSPlot.test_ts_plot_with_tz) but it's not related to what I changed.~~
- ~~Does it need a whatsnew entry?~~

### CODE REVIEW

**Comment by MJafarMashhadi** [comment]
I used `is_list_like` to check the type which does not consider strings as lists, would that be a problem? 
In the tests whereever `names` argument was passed, it was a list. The doc kind of implies it expects a list too, but the type is specified as `array-like` which I don't think is accurate because a python list is a perfectly valid argument for that but it isn't `array-like`.

**Comment by WillAyd** [comment]
Can you add a test for this? I think can go in pandas/tests/io/parser/test_common.py

**Comment by MJafarMashhadi** [comment]
Will do

**Review by jreback — APPROVED** [review]
lgtm. @gfyoung if you'd have a look.

File: pandas/tests/io/parser/test_common.py [hunk 1]
```diff
@@ -2135,3 +2135,22 @@ def test_no_header_two_extra_columns(all_parsers):
...
+    parser = all_parsers
+    msg = "Names should have consistent ordering. Consider a list instead."
+    names = "QAZ"
+    with pytest.raises(ValueError, match=msg):
+        parser.read_csv(StringIO(data), names=set(names))
+
+    ref = DataFrame(data={"Q": [1, 4, 7, 10], "A": [2, 5, 8, 11], "Z": [3, 6, 9, 12]})
+    for valid_type_converter in (list, tuple):
+        df = parser.read_csv(StringIO(data), names=valid_type_converter(names))
+        tm.assert_frame_equal(df, ref)
```
Comment (gfyoung) [inline pandas/tests/io/parser/test_common.py:2156]: Should be a separate test.  Even so, not sure I fully understand the purpose, given that we implicitly test this in all of our other `names` tests.

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -407,11 +408,15 @@ def _validate_names(names):
     Raises
     ------
     ValueError
-        If names are not unique.
+        If names are not unique or have incosistent ordering (e.g. set).
```
Comment (gfyoung) [inline pandas/io/parsers.py:411]:
```suggestion
        If names are not unique or have inconsistent ordering (e.g. set).
```

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -407,11 +408,15 @@ def _validate_names(names):
...
     """
     if names is not None:
         if len(names) != len(set(names)):
             raise ValueError("Duplicate names are not allowed.")
+        if not is_list_like(names, allow_sets=False):
+            raise ValueError(
+                "Names should have consistent ordering. Consider a list instead."
```
Comment (dsaxton) [inline pandas/io/parsers.py:418]: Wording nit, but something like "names should be an ordered collection" sounds a bit better / more correct IMHO (I wouldn't describe actually having an order as having a "consistent" order)

File: pandas/io/parsers.py (hunk 2 above)
Comment (MJafarMashhadi) [reply pandas/io/parsers.py:411]: Thanks! I missed that

File: pandas/io/parsers.py (hunk 3 above)
Comment (MJafarMashhadi) [reply pandas/io/parsers.py:418]: Thanks for your comment! The fact that there is something called "OrderedSet" suggests that the word "ordered" alone can mean the same thing that I meant by "consistent" but It's more Pythonic. I'll update the message

File: pandas/tests/io/parser/test_common.py (hunk 1 above)
Comment (MJafarMashhadi) [reply pandas/tests/io/parser/test_common.py]: Yes that makes sense. I just included this section to be more cautious; better to get a comment about the test being too much/redundant than a not-enough-tests one.

**Comment by gfyoung** [comment]
Provided tests pass, lgtm

**Comment by jreback** [comment]
thanks @MJafarMashhadi

**Pull request merged**